astropy/coordinates/astrom_provider.py

```diff
@@ -0,0 +1,220 @@
+        '''
+        Call ``erfa.apcs`` with the correct information extracted from ``frame``.
+        '''
+        pass
+
+
+class AstromProvider(BaseAstromProvider):
+    '''
+    The default provider for astrometry values.
+    A utility class to extract the necessary arguments for
+    erfa functions from frame attributes.
+    '''
+
+    @staticmethod
+    def apci(frame):
+        jd1_tt, jd2_tt = get_jd12(frame.obstime, 'tt')
+        cip = get_cip(jd1_tt, jd2_tt)
+        earth_pv, earth_heliocentric = prepare_earth_position_vel(frame.obstime)
+        return erfa.apci(jd1_tt, jd2_tt, earth_pv, earth_heliocentric, *cip)
+
+    @staticmethod
+    def apcs(frame):
+        jd1_tt, jd2_tt = get_jd12(frame.obstime, 'tt')
+        pv = pav2pv(
+            frame.obsgeoloc.get_xyz(xyz_axis=-1).value,
+            frame.obsgeovel.get_xyz(xyz_axis=-1).value
+        )
+        earth_pv, earth_heliocentric = prepare_earth_position_vel(frame.obstime)
+        return erfa.apcs(jd1_tt, jd2_tt, pv, earth_pv, earth_heliocentric)
+
+    @staticmethod
+    def apio13(frame):
+        lon, lat, height = frame.location.to_geodetic('WGS84')
+
+        jd1_utc, jd2_utc = get_jd12(frame.obstime, 'utc')
+        dut1utc = get_dut1utc(frame.obstime)
+
+        return erfa.apio13(
+            jd1_utc, jd2_utc, dut1utc,
+            lon.to_value(u.radian),
+            lat.to_value(u.radian),
+            height.to_value(u.m),
+            *get_polar_motion(frame.obstime),
+            # all below are already in correct units because they are QuantityFrameAttribues
+            frame.pressure.value,
+            frame.temperature.value,
+            frame.relative_humidity.value,
+            frame.obswl.value,
+        )
+
+
+class InterpolatingAstromProvider(AstromProvider):
+    '''
+    A provider for astrometry values that does not call erfa
+    for each individual timestamp but interpolates the linearly
+    between support points.
+
+    This can dramatically speed up coordinate transformations when
+    obstime is an array of many values (factors of 10 to > 100 depending
+    on the selected resolution and the time range of the values).
+    '''
+
+    @u.quantity_input(time_resolution=u.day)
+    def __init__(self, time_resolution):
+        if time_resolution.to_value(u.ms) < 1:
+            warnings.warn(
+                f'Using {self.__class__.__name__} with `time_resolution`'
+                ' below 1ms might not improve performance',
+                AstropyWarning
+            )
+        self.mjd_resolution = time_resolution.to_value(u.day)
+
+    def _get_support_points(self, obstime):
+        '''
+        Calculate support points for the interpolation.
+        This includes the minimum and maximum timestamp in obstime
+        to improve precision when the time resolution is wider than
+        the time range or only a single value is requested.
+        '''
+        mjd_scaled = np.ravel(obstime.mjd / self. mjd_resolution)
```

Comment:
With float64, the minimal `mjd_resolution` will be of the order of microseconds before numerical accuracy issues kick in (please confirm). I wonder, if one should restrict the range of mjd resolutions the user is allowed to set?

Reply:
You mean raising an error instead of just a warning?

Reply:
The thing is, when float64 is not sufficiently accurate for the `mjd/res` operation (float64 is already problematic for mjd anyway, if high time resolution is needed), the result could simply be wrong. This is perhaps reason to raise an error, but I don't have strong feelings, if you prefer a warning. Nevertheless, this scenario should raise a different warning/error than the one, which is already included (about the performance implications).

Reply:
Yes, should be:

```
>>> import numpy as np
>>> import astropy.units as u
>>> from astropy.time import Time
>>> (Time.now().mjd * np.finfo(np.float64).eps * u.day).to(u.us)
>>> <Quantity 1.13335838 us>
```

So warning for < 1ms and exception for < 10 μs?

Reply:
I'd stick with warning, but expand the docstring and refer to that in the warning.

Reply:
p.s. Basically, if we start going the warning route, we should also warn against overly long intervals - I feel this is specialized enough that users should really read the documentation. 

A propos documentation, this obviously deserves a nice entry in the performance section!

Reply:
Ok, I'll raise another warning for resolution < 10µs and start working on an example for the performance section.

Reply:
Concerning overly long, that really depends on the needed accuracy. People will anyways have to make sure by doing comparisons and checks, that the interpolation version delivers the precision they need. 

In my test, even intervals as long as 6 hours still produced < mas resolution.

Reply:
Agreed - this would seem to be a question of documentation.